File: src/main/java/com/busted_moments/client/models/territory/TerritoryModel.java
package com.busted_moments.client.models.territory;

import com.busted_moments.client.events.mc.MinecraftStartupEvent;
import com.busted_moments.client.models.territory.events.MapUpdateEvent;
import com.busted_moments.client.models.territory.events.TerritoryCapturedEvent;
import com.busted_moments.core.Model;
import com.busted_moments.core.http.requests.mapstate.MapState;
import com.busted_moments.core.http.requests.mapstate.Territory;
import com.busted_moments.core.heartbeat.Heartbeat;
import com.busted_moments.core.json.Json;
import com.busted_moments.core.time.ChronoUnit;
import com.wynntils.core.text.PartStyle;
import com.wynntils.handlers.chat.event.ChatMessageReceivedEvent;
import com.wynntils.mc.event.TickEvent;
import net.minecraftforge.eventbus.api.EventPriority;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static com.wynntils.utils.mc.McUtils.player;

@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
public class TerritoryModel extends Model {
   private final static Pattern TERRITORY_CAPTURED_PATTERN = Pattern.compile("^\\[WAR] \\[(?<guild>.+)] captured the territory (?<territory>.+)\\.");
   private final static Pattern TERRITORY_CONTROL_PATTERN = Pattern.compile("^\\[INFO] \\[(?<guild>.+)] has taken control of (?<territory>.+)!");

   @SuppressWarnings("unused")
   private Socket ACTIVE_SOCKET;

   private MapState LATEST_TERRITORIES;

   @Instance
   private static TerritoryModel THIS;

   private static Optional<Territory> CURRENT_TERRITORY = Optional.empty();

   @SubscribeEvent
   public void onMessage(ChatMessageReceivedEvent event) {
      Matcher matcher = event.getOriginalStyledText().getMatcher(TERRITORY_CAPTURED_PATTERN, PartStyle.StyleType.NONE);
      if (matcher.matches()) {
         new TerritoryCapturedEvent(matcher.group("territory"), matcher.group("guild")).post();
         return;
      }

      matcher = event.getOriginalStyledText().getMatcher(TERRITORY_CONTROL_PATTERN, PartStyle.StyleType.NONE);
      if (matcher.matches())
         new TerritoryCapturedEvent(matcher.group("territory"), matcher.group("guild")).post();
   }


   @SubscribeEvent
   public static void onGameStart(MinecraftStartupEvent event) {
      new Territory.Request().thenAccept(optional -> {
         new MapUpdateEvent(optional.orElse(MapState.empty())).post();

         THIS.ACTIVE_SOCKET = THIS.new Socket();
      });
   }

   @SubscribeEvent(priority = EventPriority.LOWEST)
   public void onMapUpdate(MapUpdateEvent event) {
      if (LATEST_TERRITORIES != null)
         event.getState().forEach(territory -> {
            if (LATEST_TERRITORIES.contains(territory) && !LATEST_TERRITORIES.get(territory).getOwner().equals(territory.getOwner()))
               new TerritoryCapturedEvent(territory.getName(), territory.getOwner().prefix());
         });

      LATEST_TERRITORIES = event.getState();
   }

   @SubscribeEvent
   public void onTick(TickEvent event) {
      if (LATEST_TERRITORIES != null) {
         CURRENT_TERRITORY = LATEST_TERRITORIES.stream()
                 .filter(territory -> territory.getLocation().isInside(player().position()))
                 .findFirst();
      }
   }

   public static MapState getTerritoryList() {
      return THIS.LATEST_TERRITORIES;
   }

   public static Optional<Territory> getCurrentTerritory() {
      return CURRENT_TERRITORY;
   }

   protected class Socket extends WebSocketClient {
      private static final URI SOCKET_URI = URI.create("wss://thesimpleones.net/war");

      public Socket() {
         super(SOCKET_URI);

         connect();
      }

      @Override
      public void onOpen(ServerHandshake handshake) {

      }

      @Override
      public void onMessage(String message) {
         Json.tryParse(message).ifPresent(json -> {
            if (json.getString("type").equals("MapUpdate")) {
               new MapUpdateEvent(json.getJson("map").wrap(MapState::new)).post();
            }
         });
      }

      @Override
      public void onClose(int code, String reason, boolean remote) {
         LOGGER.warn("Territory socket has disconnected with code {} (reason={}, remote={})", code, reason, reason);

         Heartbeat.schedule(() -> ACTIVE_SOCKET = new Socket(), 10, ChronoUnit.SECONDS);
      }

      @Override
      public void onError(Exception ex) {
         LOGGER.error("Error in socket", ex);
      }
   }
}

File: src/main/kotlin/com/busted_moments/client/models/territories/TerritoryModel.kt
package com.busted_moments.client.models.territories

import com.busted_moments.buster.api.Territory
import com.busted_moments.client.buster.TerritoryList
import com.busted_moments.client.framework.events.Subscribe
import com.busted_moments.client.models.territories.TerritoryModel.clean
import com.wynntils.mc.event.TickEvent
import com.wynntils.utils.mc.McUtils.player
import java.util.regex.Pattern

private val CLEANER = Pattern.compile("[^\\w\\s]")

object TerritoryModel {
    var inside: Territory? = null
        private set

    @Subscribe
    fun TickEvent.on() {
        inside = TerritoryList.firstOrNull {
            it.location.contains(
                player().x.toInt(),
                player().z.toInt()
            )
        }
    }

    fun Territory.Location.contains(x: Int, z: Int): Boolean {
        return start.x <= x && end.x >= x && start.z <= z && end.z >= z
    }

    private fun String.clean(): String =
        CLEANER.matcher(this).replaceAll("").lowercase()

    fun getAcronym(string: String) =
        string.split(' ', '-', ignoreCase = true)
        .asSequence()
        .map { if (it.isBlank()) "" else it[0] }
        .joinToString(separator = "")

    fun matches(input: String, search: String): Boolean {
        if (search.isBlank())
            return true

        val cleaned = input.clean().also { clean ->
            val acronym = clean.split(' ', '-', ignoreCase = true)
                .asSequence()
                .map { if (it.isBlank()) "" else it[0] }
                .joinToString(separator = "")

            if (acronym.startsWith(search, ignoreCase = true))
                return true
        }.split(' ')

        val parts = search.clean().split(' ')

        if (parts.size > cleaned.size)
            return false

        var offset: Int = -1

        for (i in cleaned.indices) {
            if (offset == -1 && cleaned[i].contains(parts[0], ignoreCase = true))
                offset = i

            if (offset == -1)
                continue

            val index = i - offset
            if (index >= parts.size)
                break

            if (!cleaned[i].contains(parts[index], ignoreCase = true))
                offset = -1
        }

        return offset != -1 && (parts.size + offset) <= cleaned.size
    }
}